File: apprefurbished/src/main/java/com/androidcollider/easyfin/transaction_categories/nested/TransactionCategoriesNestedMVP.java
package com.androidcollider.easyfin.transaction_categories.nested;

import android.content.res.TypedArray;
import android.os.Bundle;
import android.support.annotation.Nullable;

import com.androidcollider.easyfin.common.models.TransactionCategory;

import java.util.List;

import io.reactivex.Flowable;

/**
 * @author Ihor Bilous
 */

interface TransactionCategoriesNestedMVP {

    interface Model {

        Flowable<List<TransactionCategory>> getTransactionCategories(boolean isExpense);

        Flowable<TransactionCategory> updateTransactionCategory(TransactionCategory transactionCategory, boolean isExpense);

        Flowable<Boolean> deleteTransactionCategory(int id, boolean isExpense);
    }

    interface View {

        void setTransactionCategoryList(List<TransactionCategory> transactionCategoryList, TypedArray iconsArray);

        void showMessage(String message);

        void shakeDialogUpdateTransactionCategoryField();

        void dismissDialogUpdateTransactionCategory();

        void handleTransactionCategoryUpdated();

        void deleteTransactionCategory();
    }

    interface Presenter {

        void setView(@Nullable TransactionCategoriesNestedMVP.View view);

        void loadData();

        void loadTransactionCategories();

        void setArguments(Bundle args);

        void updateTransactionCategory(int id, String name);

        void deleteTransactionCategoryById(int id);

        String getCategoryNameById(int id);
    }
}

File: apprefurbished/src/main/java/com/androidcollider/easyfin/transaction_categories/nested/TransactionCategoriesNestedMVP.kt
package com.androidcollider.easyfin.transaction_categories.nested

import android.content.res.TypedArray
import android.os.Bundle
import com.androidcollider.easyfin.common.models.TransactionCategory
import io.reactivex.rxjava3.core.Single

/**
 * @author Ihor Bilous
 */
interface TransactionCategoriesNestedMVP {
    interface Model {
        fun getTransactionCategories(isExpense: Boolean): Single<List<TransactionCategory>>
        fun updateTransactionCategory(
            transactionCategory: TransactionCategory,
            isExpense: Boolean
        ): Single<TransactionCategory>

        fun deleteTransactionCategory(id: Int, isExpense: Boolean): Single<Boolean>
    }

    interface View {
        fun setTransactionCategoryList(
            transactionCategoryList: List<TransactionCategory>,
            iconsArray: TypedArray
        )

        fun showMessage(message: String)
        fun shakeDialogUpdateTransactionCategoryField()
        fun dismissDialogUpdateTransactionCategory()
        fun handleTransactionCategoryUpdated()
        fun deleteTransactionCategory()
    }

    interface Presenter {
        fun setView(view: View?)
        fun loadData()
        fun loadTransactionCategories()
        fun setArguments(args: Bundle?)
        fun updateTransactionCategory(id: Int, name: String)
        fun deleteTransactionCategoryById(id: Int)
        fun getCategoryNameById(id: Int): String
    }
}